src/main/java/com/artipie/asto/Key.java

```diff
@@ -83,6 +83,21 @@ public final String string() {
         public Optional<Key> parent() {
             return this.origin.parent();
         }
+
+        @Override
+        public boolean equals(final Object another) {
+            return this.origin.equals(another);
```

Comment:
@Sammers21 this implementation is not correct, according to `equals` specification of `java.lang.Object` :
 1. It should be "reflexive", but now `key.equals(key)` will fall into stack overflow error
 2. It should be "symmetric", but it is not: `wrap.equals(origin) != origin.equals(wrap)`

[See the docs](https://docs.oracle.com/en/java/javase/11/docs/api/java.base/java/lang/Object.html#equals(java.lang.Object)):
>  The equals method implements an equivalence relation on non-null object references:
>
>    It is reflexive: for any non-null reference value x, x.equals(x) should return true.
>    It is symmetric: for any non-null reference values x and y, x.equals(y) should return true if and only if y.equals(x) returns true.
>    It is transitive: for any non-null reference values x, y, and z, if x.equals(y) returns true and y.equals(z) returns true, then x.equals(z) should return true.
>    It is consistent: for any non-null reference values x and y, multiple invocations of x.equals(y) consistently return true or consistently return false, provided no information used in equals comparisons on the objects is modified.
>    For any non-null reference value x, x.equals(null) should return false. 
>
> The equals method for class Object implements the most discriminating possible equivalence relation on objects;  that is, for any non-null reference values x and y, this method returns true if and only if x and y refer to the same object (x == y has the value true).

Note that it is generally necessary to override the hashCode method whenever this method is overridden, so as to maintain the general contract for the hashCode method, which states that equal objects must have equal hash codes.